lib/charms/synapse/v0/matrix_auth.py

```diff
@@ -0,0 +1,512 @@
+
+        Args:
+            model: the Juju model.
+            relation: relation to grant access to the secrets to.
+
+        Returns:
+            Dict containing the representation.
+        """
+        self.set_registration_id(model, relation)
+        dumped_model = self.model_dump(exclude_unset=True)
+        dumped_data = {
+            "registration_secret_id": dumped_model["registration_secret_id"],
+        }
+        return dumped_data
+
+    @classmethod
+    def from_relation(cls, model: ops.Model, relation: ops.Relation) -> "MatrixAuthRequirerData":
+        """Get a MatrixAuthRequirerData from the relation data.
+
+        Args:
+            model: the Juju model.
+            relation: the relation.
+
+        Returns:
+            the relation data and the processed entries for it.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        try:
+            app = cast(ops.Application, relation.app)
+            relation_data = relation.data[app]
+            registration_secret_id = (
+                (relation_data["registration_secret_id"])
+                if "registration_secret_id" in relation_data
+                else None
+            )
+            registration = MatrixAuthRequirerData.get_registration(model, registration_secret_id)
+            return MatrixAuthRequirerData(
+                registration=registration,
+            )
+
+        except json.JSONDecodeError as ex:
+            logger.warning("Invalid relation data %s", ex)
+            raise ValueError from ex
+
+
+#### Events ####
+class MatrixAuthRequestProcessed(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is processed."""
+
+    def get_matrix_auth_provider_relation_data(self) -> MatrixAuthProviderData:
+        """Get a MatrixAuthProviderData for the relation data.
+
+        Returns:
+            the MatrixAuthProviderData for the relation data.
+        """
+        return MatrixAuthProviderData.from_relation(self.model, self.relation)
+
+
+class MatrixAuthRequestReceived(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is made."""
+
+
+class MatrixAuthRequiresEvents(ops.CharmEvents):
+    """MatrixAuth requirer events.
+
+    This class defines the events that a MatrixAuth requirer can emit.
+
+    Attributes:
+        matrix_auth_request_processed: the MatrixAuthRequestProcessed.
+    """
+
+    matrix_auth_request_processed = ops.EventSource(MatrixAuthRequestProcessed)
+
+
+class MatrixAuthProvidesEvents(ops.CharmEvents):
+    """MatrixAuth provider events.
+
+    This class defines the events that a MatrixAuth provider can emit.
+
+    Attributes:
+        matrix_auth_request_received: the MatrixAuthRequestReceived.
+    """
+
+    matrix_auth_request_received = ops.EventSource(MatrixAuthRequestReceived)
+
+
+#### Requires and Provides ####
+class MatrixAuthRequires(ops.Object):
```

Comment:
Nitpick: Keep the same order as for the pydantic model, in the pydantic model the "Provider" is before the "Requirer"

Reply:
Closing this one since this library is hosted in the synapse repo.